### PR TITLE
Adding test

### PR DESCRIPTION
**Pull request by pinghaoluo**

### CODE REVIEW

File: src/service/service_layer.cc [hunk 1]
```diff
@@ -100,6 +105,35 @@ void ServiceLayer::ReadDfs(const std::string& key_base,
   ReadDfs(key_base, chirps, counter + 1);
 }
 
+std::vector<ChirpObj> ServiceLayer::HashTag(const std::string& uname,const std::string& hash_tag) {
```
Comment (allenyhu) [inline src/service/service_layer.cc]: Why do you change Monitor()?

File: src/service/service_layer.cc [hunk 2]
```diff
@@ -100,6 +105,35 @@ void ServiceLayer::ReadDfs(const std::string& key_base,
...
+  //getting all users
+  std::vector<std::string> users = ds_.Get(AllUserKey_); 
+  for (const std::string& u : users) {
+    PutMonitorKey(uname, u);
+  }
+
+  std::vector<ChirpObj> chirps;
+  std::string monitor_check_base = uname + kMonitorCheckKey_;
```
Comment (allenyhu) [inline src/service/service_layer.cc:116]: Why do you use monitor bookkeeping info?

File: src/service/service_layer.cc [hunk 3]
```diff
@@ -125,6 +159,50 @@ std::vector<ChirpObj> ServiceLayer::Monitor(const std::string& uname) {
...
+      for(auto p:positions){
+        std::string sub = text.substr(p);
+        std::size_t pos1 = sub.find(" ");
+        std::string tag = sub.substr (1,pos1);
+        //trim the tags with white space
+        // tag.erase( std::remove_if( tag.begin(), tag.end(), ::isspace ), tag.end() );
+        if(tag == hash_tag){
+          // std::cout<<tag<<std::endl;
+          // std::cout<<hash_tag<<std::endl;
+          contain = true;
```
Comment (allenyhu) [inline src/service/service_layer.cc]: why not just return after this?

File: src/service/service_layer.cc [hunk 4]
```diff
@@ -125,6 +159,50 @@ std::vector<ChirpObj> ServiceLayer::Monitor(const std::string& uname) {
...
+        }
+      }
+      //if doesnt contain 
+      if(contain == false){
+        it = chirps.erase(it);
+        continue;
```
Comment (allenyhu) [inline src/service/service_layer.cc]: is this necessary?

File: src/service/service_layer.h [hunk 5]
```diff
@@ -45,7 +45,16 @@ class ServiceLayer {
...
   std::vector<ChirpObj> Monitor(const std::string& uname);
 
+// Streams Chirps from all followed users
+  // @uname: the user requesting the monitor
+  // @ret: the Chirps of all followed users
+  std::vector<ChirpObj> HashTag(const std::string& uname,const std::string& hash_tag);
+
  private:
+  //Store all users
+  const std::string AllUserKey_ = "-all-user-key";
```
Comment (allenyhu) [inline src/service/service_layer.h:55]: rename kAllUserKey_

File: src/service/service_layer_tests.cc [hunk 6]
```diff
@@ -324,6 +324,79 @@ TEST(ServiceLayerMonitor, BaseMonitorOnlyOneFollowed) {
...
+
+  check = s.HashTag("user1","hi");
+  ASSERT_EQ(1, check.size());
+  EXPECT_EQ(c1.to_string(), check[0].to_string());
+}
+
+// Tests HashTag function for ServiceLayer
+// Due to lack of GRPC, will return vector of store stream chirps
+// new chirps Should only return vector is it contains the hash tag
+TEST(ServiceLayerStream, StramingChirpWithNoHashTag) {
```
Comment (allenyhu) [inline src/service/service_layer_tests.cc]: StreamingChirpWithNoHashTag

File: src/service/service_layer_tests.cc [hunk 7]
```diff
@@ -324,6 +324,79 @@ TEST(ServiceLayerMonitor, BaseMonitorOnlyOneFollowed) {
...
+  ASSERT_FALSE(c2.id().empty());
+
+  check = s.HashTag("user1","hi");
+  EXPECT_TRUE(check.empty());
+}
+
+// Tests HashTag function for ServiceLayer
+// Due to lack of GRPC, will return vector of store stream chirps
+// new chirps Should only return vector is it contains the hash tag
+TEST(ServiceLayerMonitor, StramingWithMultipleUser) {
```
Comment (allenyhu) [inline src/service/service_layer_tests.cc]: StreamingWithMultipleUser

File: src/service/service_layer.h [hunk 8]
```diff
@@ -71,6 +80,11 @@ class ServiceLayer {
   // @ret: ChirpObj constructed from `chirp`
   ChirpObj ParseChirpString(const std::string& chirp);
 
+  // Helper function to parse data from DataStore
```
Comment (barath) [inline src/service/service_layer.h:83]: This comment should specify what `chirps` contains and what the return vector contains in relation to `chirps`.

Also, you shouldn't pass in a non-const reference.

File: src/service/service_layer.cc [hunk 9]
```diff
@@ -125,6 +159,50 @@ std::vector<ChirpObj> ServiceLayer::Monitor(const std::string& uname) {
...
+      }else{
+        it++;
+      }
+    }
+  return chirps;
```
Comment (barath) [inline src/service/service_layer.cc]: This is a bit odd -- if you are returning chirps you shouldn't be modifying it.

File: src/service/service_layer.cc [hunk 10]
```diff
@@ -125,6 +159,50 @@ std::vector<ChirpObj> ServiceLayer::Monitor(const std::string& uname) {
   return chirps;
 }
 
+std::vector<ChirpObj> ServiceLayer::CheckTag(std::vector<ChirpObj> &chirps,const std::string& hash_tag){
+  for(std::vector<ChirpObj>::iterator it = chirps.begin(); it != chirps.end();) {
```
Comment (barath) [inline src/service/service_layer.cc]: I would recommend restructuring this loop.  Instead, you can write a helper that checks if a hashtag is in a single ChirpObj, and then call that on each, and add the ones that match to a return vector.

File: src/service/service_layer.cc (hunk 4 above)
Comment (pinghaoluo) [reply src/service/service_layer.cc]: Thanks, I will fix this, I will write a function to check if a chirp contain hashtag

File: src/service/service_layer.cc (hunk 2 above)
Comment (pinghaoluo) [reply src/service/service_layer.cc:116]: I was little confused about the Key and Value you stored in KV store for monitor. So I keep using the previous code in Monitor for Stream. I am going to fix it.

Comment (pinghaoluo) [reply src/service/service_layer.cc:116]: Do you have a function checking if a user is register?

Comment (allenyhu) [reply src/service/service_layer.cc:116]: you can just check if the username exists as a key in the store

**Pull request merged**